R/test.R

```diff
@@ -43,99 +45,126 @@ isShinyTest <- function(text){
+#' | `result` | any or `NA` | The return value of the runner |
 #'
 #' @details Historically, [shinytest](https://rstudio.github.io/shinytest/)
-#'   recommended placing tests at the top-level of the `tests/` directory. In
-#'   order to support that model, `testApp` first checks to see if the `.R`
-#'   files in the `tests/` directory are all shinytests; if so, just calls out
-#'   to [shinytest::testApp()].
+#'   recommended placing tests at the top-level of the `tests/` directory.
+#'   This older folder structure is not supported by runTests.
+#'   Please see [shinyAppTemplate()] for more details.
 #' @export
-runTests <- function(appDir=".", filter=NULL){
+runTests <- function(
+  appDir = ".",
+  filter = NULL,
+  assert = TRUE,
+  envir = globalenv()
+) {
+  # make sure shiny is available to use without `::`
   require(shiny)
 
   testsDir <- file.path(appDir, "tests")
-  if (!dirExists(testsDir)){
+  if (!dirExists(testsDir)) {
     stop("No tests directory found: ", testsDir)
   }
   runners <- list.files(testsDir, pattern="\\.r$", ignore.case = TRUE)
 
-  if (length(runners) == 0){
+  if (length(runners) == 0) {
     message("No test runners found in ", testsDir)
-    return(result_row(character(0), logical(0), list(), list()))
+    return(result_row(character(0), logical(0), list()))
   }
 
-  if (!is.null(filter)){
+  if (!is.null(filter)) {
     runners <- runners[grepl(filter, runners)]
   }
-  if (length(runners) == 0){
+  if (length(runners) == 0) {
     stop("No test runners matched the given filter: '", filter, "'")
   }
 
-  # Inspect each runner to see if it appears to be a shinytest
-  isST <- vapply(runners, function(r){
-    text <- readLines(file.path(testsDir, r), warn = FALSE)
-    isShinyTest(text)
-  }, logical(1))
-
   # See the @details section of the runTests() docs above for why this branch exists.
-  if (all(isST)){
-    # just call out to shinytest
-    # We don't need to message/warn here since shinytest already does it.
-    if (!requireNamespace("shinytest", quietly=TRUE) ){
-      stop("It appears that the .R files in ", testsDir,
-           " are all shinytests, but shinytest is not installed.")
-    }
-
-    if (!getOption("shiny.autoload.r", TRUE)) {
-      warning("You've disabled `shiny.autoload.r` via an option but this is not passed through to shinytest. Consider using a _disable_autoload.R file as described at https://rstd.io/shiny-autoload")
-    }
-
-    return(do.call(rbind, lapply(shinytest::testApp(appDir)[["results"]], function(r) {
-      error <- if (r[["pass"]]) NA else simpleError("Unknown shinytest error")
-      result_row(r[["name"]], r[["pass"]], list(NA), list(error))
-    })))
+  if (isLegacyShinyTest(file.path(testsDir, runners))) {
+    stop(
+      "It appears that the .R files in ", testsDir, " are all shinytests.",
+      "\nThis is not supported by `shiny::runTests()`.",
+      " Please see `?shiny::shinyAppTemplate` on how to structure your shinytest tests."
```

Comment:
"Please see `?shiny::migrateLegacyShinytest` to migrate your shinytest file structure to the new format. Also see `?shiny::shinyAppTemplate` for an example of the new format."